src/utils/pgwire/src/pg_protocol.rs

```diff
@@ -620,21 +625,22 @@ impl<S> PgStream<S>
 where
     S: AsyncWrite + AsyncRead + Unpin,
 {
-    async fn ssl(&mut self, ssl_ctx: &SslContextRef) -> PgStream<SslStream<S>> {
+    async fn ssl(&mut self, ssl_ctx: &SslContextRef) -> PsqlResult<PgStream<SslStream<S>>> {
         // Note: Currently we take the ownership of previous Tcp Stream and then turn into a
         // SslStream. Later we can avoid storing stream inside PgProtocol to do this more
         // fluently.
         let stream = self.stream.take().unwrap();
         let ssl = openssl::ssl::Ssl::new(ssl_ctx).unwrap();
         let mut stream = tokio_openssl::SslStream::new(ssl, stream).unwrap();
         if let Err(e) = Pin::new(&mut stream).accept().await {
-            panic!("Unable to set up a ssl connection, reason: {}", e);
+            warn!("Unable to set up a ssl connection, reason: {}", e);
```

Comment:
You only get to this code after terminating the connection.

Reply:
I tried a similar direction with 

```rust 
        if let Err(e) = Pin::new(&mut stream).accept().await {
            tracing::error!("Unable to establish the SSL connection. Error was '{}'", e);
            return PgStream {
                stream: None,
                write_buf: BytesMut::with_capacity(0),
            };
        }
```
But it went nowhere

Reply:
Is there the possibility to set some timeout for `Pin::new(&mut stream).accept().await`. I think if we could do that your code this may work.

Reply:
No, i need to reject this connection. I will finish this later.

Reply:
Expected behaviour is you never login to the pg server but server do not panic (only warning)

Reply:
Yes. I was also surprised by that. I would have expected that the server rejects the connection if SSL fails.